### ISSUE TITLE
Build failing with `spawn ENOMEM` / `spawn E2BIG`

### ISSUE THREAD

**Issue by znarf** [zeit/next.js#6353]
# Bug report

## Describe the bug

In some circumstances, `next build` is failing in CI environment (CircleCI) with the exception `Callback was already called'. 

There is a random factor and it's not happening 100% of the time, but more than 50% at the moment. 

This is triggered from the `neo-async` package which is used by `webpack`:

```
/home/circleci/project/node_modules/neo-async/async.js:14
    throw new Error('Callback was already called.');
    ^

Error: Callback was already called.
    at throwError (/home/circleci/project/node_modules/neo-async/async.js:14:11)
    at /home/circleci/project/node_modules/neo-async/async.js:2805:7
    at afterBuild (/home/circleci/project/node_modules/webpack/lib/Compilation.js:859:17)
    at processTicksAndRejections (internal/process/next_tick.js:74:9)
```

## Example

See for example: https://circleci.com/gh/opencollective/opencollective-frontend/5926

## Version

This is a new issue after upgrading to next 8.0.x, we never had it with next 7.0.x.

**Comment by timneutkens** [comment]
From checking Google it seems like the culprit is installing webpack on top of webpack, eg what you're doing here: https://github.com/opencollective/opencollective-frontend/blob/master/package.json#L99

**Comment by timneutkens** [comment]
I still have no clue why people install webpack as well as Next.js, could you provide more context into this?

**Comment by timneutkens** [comment]
We're currently working on bundling webpack into Next.js using ncc: #6347 which would probably solve the inconsistency issue.

**Comment by znarf** [comment]
@timneutkens Because we are requiring/importing webpack in `next.config.js` I thought this was a good idea to explicitly import it in package.json.

> I still have no clue why people install webpack as well as Next.js, could you provide more context into this?

Some linters are complaining about that (missing import) and that might be why people actually do this. This was a motivating factor for me.

I was also aware of the potential conflict and this is why I explicitly picked the version matching next.js. But it seems that's still a bad idea.

Thank you for the feedback, I will try removing webpack from our package.json. I will confirm here if this is fixing it.

**Comment by timneutkens** [comment]
I think we should provide `webpack` as part of the options object (second argument) 🤔

```
webpack(config, {webpack}) {
  config.plugins.push(webpack.DefinePlugin({/*plugin configuration*/}))
}
```

That would probably make sense 🤔

**Comment by znarf** [comment]
@timneutkens For `webpack` as part of the options object, I think that's a great idea!

For the issue I reported, the proposed fix is unfortunately not working:
https://github.com/opencollective/opencollective-frontend/pull/1422
https://circleci.com/gh/opencollective/opencollective-frontend/5165

**Comment by wtgtybhertgeghgtwtg** [comment]
> Some linters are complaining about that (missing import) and that might be why people actually do this.

Installing some of the add-on's (i.e. `@zeit/next-css`) without `webpack` gives warnings, as well.
```
warning "@zeit/next-css > css-loader@1.0.0" has unmet peer dependency "webpack@^4.0.0".
warning "@zeit/next-css > mini-css-extract-plugin@0.4.3" has unmet peer dependency "webpack@^4.4.0".
```

**Comment by timneutkens** [comment]
It sounds like a pretty horrible idea for those to show webpack installation warnings though wrt tooling etc 😰 We might be able to solve that by `ncc`ing (github.com/zeit/ncc) the deps.

**Comment by znarf** [comment]
@timneutkens I created an a new branch to reproduce the issue (without the webpack duplication).

https://github.com/opencollective/opencollective-frontend/pull/1442
https://circleci.com/gh/opencollective/opencollective-frontend/5930

What can I do to help troubleshot?

**Comment by ijjk** [comment]
@znarf I just pulled down your branch from https://github.com/opencollective/opencollective-frontend/pull/1442 and when running with CircleCI locally using their cli tool the `build` job seems to work fine. It might have something to do with your npm caching set up in CircleCI.

**Comment by znarf** [comment]
@ijjk Why would caching the `.npm` folder would trigger a problem in the build?

I acknowledge it's working with `circleci build` locally, I guess there are significant differences between that and the hosted CircleCI.

I created a dedicated branch disabling the cache https://github.com/opencollective/opencollective-frontend/tree/package/next8-no-cache and it is still failing https://circleci.com/gh/opencollective/opencollective-frontend/8314

**Comment by ijjk** [comment]
@znarf interesting I was thinking the main difference between the local CircleCI and the one that was failing was the cache but there must be something else affecting the build then. 🤔

**Comment by ijjk** [comment]
@znarf Do you mind if I open a PR on your project to test in CircleCI since I can't reproduce it locally?

**Comment by znarf** [comment]
@ijjk please do! I invited you to the repository, you can push your branch directly on `opencollective-frontend`.

**Comment by ijjk** [comment]
@znarf I just submitted a PR and it is passing CircleCI now. It appears this was caused from pre-compiling `src/pages` and `src/components` before building with Next.js. 

I'm gonna close this issue now, feel free to comment if you have additional trouble with this issue.

**Comment by znarf** [comment]
@ijjk Thank you for that! I did know that pre-compiling `src/pages` and `src/components` was not a great idea, but could not link that to the issue. I thought I already tried that, but maybe I'm wrong. I'll run CircleCI a couple of times to ensure this is passing consistently.

Unfortunately had an error at the first two attempts:
https://circleci.com/gh/opencollective/opencollective-frontend/8457
https://circleci.com/gh/opencollective/opencollective-frontend/8463
```
Error: spawn E2BIG
    at ChildProcess.spawn (internal/child_process.js:372:11)
    at spawn (child_process.js:559:9)
    at Object.fork (child_process.js:113:10)
    at fork (/home/circleci/project/node_modules/worker-farm/lib/fork.js:17:36)
    at Farm.startChild (/home/circleci/project/node_modules/worker-farm/lib/farm.js:106:16)
    at Farm.processQueue (/home/circleci/project/node_modules/worker-farm/lib/farm.js:279:10)
    at Farm.<anonymous> (/home/circleci/project/node_modules/worker-farm/lib/farm.js:97:21)
    at listOnTimeout (timers.js:324:15)
    at processTimers (timers.js:268:5)
```

**Comment by ijjk** [comment]
@znarf This is interesting 🤔I just pushed up another commit let's see how that one does

**Comment by ijjk** [comment]
This issue seems to be finicky, re-opening so others can investigate also

**Comment by adam187** [comment]
I can confirm, have a very similar case. 

Sometimes it works, sometimes I'm getting `spawn E2BIG` error. 

And sometimes with:
```
Error: ENOMEM: not enough memory, read
    at Object.readSync (fs.js:494:3)
    at tryReadSync (fs.js:333:20)
    at Object.readFileSync (fs.js:362:19)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:699:20)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at Object.<anonymous> (node_modules/terser-webpack-plugin/dist/minify.js:8:15)
internal/child_process.js:366
    throw errnoException(err, 'spawn');
```

and sometimes i'm getting time out. 

Locally build works, also when a build is run inside docker container on CircleCI it seems to work without issues.

**Comment by znarf** [comment]
I decided to modify our repository with a change that was recommended by @ijjk: https://github.com/opencollective/opencollective-frontend/pull/1517

As expected, the issue morphed into a new version, throwing `spawn ENOMEM` or `spawn E2BIG`. Same one as @adam187.

On our side, we're gonna try to fix this new form, not the older one.

**Comment by ijjk** [comment]
@znarf It looks like in CircleCI the `os.cpus()` returns all the CPUs and not the limited amount for the current run. This explains why it was working in local CircleCI but not in their environment. This has been updated and we also added passing `webpack` to the `config.webpack` function.

The draft PR I opened should allow you to update to v8 after we release a new canary with this fix. There is one e2e test failing on that PR and I'm not sure if that is due to Next 8 or not. If it is I can look into it further.
